lib/index.js

```diff
@@ -2,3 +2,163 @@
  * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
  */
 'use strict';
+const {generateId} = require('bnid');
+const {snakeCase} = require('snake-case');
+const {
+  InvalidRequest, AccessDenied
+} = require('@interop/oauth2-errors');
+
+const {ClientMetadata, FORBIDDEN_FIELDS} = require('./ClientMetadata');
+
+function handleClientRegistration({authentication, register,
+  defaults = {}, allowClientProvidedId = false, logger}) {
+  return async (req, res) => {
+    try {
+      const initialAccessToken = await _authenticate({
+        req, authentication, logger});
+
+      const {body: registrationBody} = req;
+
+      const registration = _parseRegistration({registrationBody, defaults});
+
+      const result = registration.validateNew();
+      if(!result.valid) {
+        throw result.error;
+      }
+
+      // client can be optionally overriden by the register() callback
+      await _ensureClientId({registration, allowClientProvidedId});
+      await _generateClientSecret({registration, defaults});
+
+      // registration.registration_access_token = '...';
+      // registration.registration_client_uri = (new URL('/register/' +
+      //   encodeURIComponent(registration.client_id), baseUrl)).toString();
+
+      // Epoch time
+      registration.client_id_issued_at = Math.floor(Date.now() / 1000);
+
+      let client;
+      try {
+        client = await register({
+          registration, credentials: {initialAccessToken}
+        });
+      } catch(e) {
+        if(logger) {
+          logger.error(e);
+        }
+        throw new InvalidRequest({
+          description: 'Error registering client.'
+        });
+      }
+
+      _respond({client, res});
+    } catch(error) {
+      if(logger) {
+        logger.error(error);
+      }
+      error.respond(res);
+    }
+  };
+}
+
+function _respond({client, res}) {
+  res.set({
+    'Cache-Control': 'no-store',
+    Pragma: 'no-cache'
+  });
+
+  res.status(201).json(client);
+}
+
+function _parseRegistration({registrationBody, defaults}) {
+  if(!registrationBody) {
+    throw new InvalidRequest({
+      error: 'invalid_request',
+      error_description: 'Missing registration request body.'
+    });
+  }
+
+  const hit = FORBIDDEN_FIELDS.find(prop =>
```

Comment:
i think inverting this an iterating over the registration body properties and making FORBIDDEN_FIELDS a SET would be in improvement.

Reply:
Sure, using a Set sounds reasonable.

Reply:
Refactored to use Set, inverted, added test.